pkg/controller/logstorage/linseed/esgateway.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/tigera/operator/pkg/controller/utils"
+	"github.com/tigera/operator/pkg/controller/utils/imageset"
+	"github.com/tigera/operator/pkg/render"
+	"github.com/tigera/operator/pkg/render/common/secret"
+	"github.com/tigera/operator/pkg/render/logstorage/esgateway"
+	"github.com/tigera/operator/pkg/tls/certificatemanagement"
+)
+
+func (r *LinseedSubController) createESGateway(
+	ctx context.Context,
+	request octrl.Request,
+	install *operatorv1.InstallationSpec,
+	variant operatorv1.ProductVariant,
+	pullSecrets []*corev1.Secret,
+	hdler utils.ComponentHandler,
+	reqLogger logr.Logger,
+	usePSP bool,
+) error {
+	// Create a trusted bundle to pass to the render pacakge. The actual contents of this bundle don't matter - the ConfigMap
+	// itself will be managed by the Secret controller. But, we need an interface to use as an argument to render in order
+	// to configure volume mounts properly.
+	trustedBundle := certificatemanagement.CreateTrustedBundle()
+
+	// Get the ES admin user secret. This is provisioned by the ECK operator as part of installing Elasticsearch,
+	// and so may not be immediately available.
+	esAdminUserSecret, err := utils.GetSecret(ctx, r.client, render.ElasticsearchAdminUserSecret, render.ElasticsearchNamespace)
+	if err != nil {
+		reqLogger.Error(err, "failed to get Elasticsearch admin user secret")
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Failed to get Elasticsearch admin user secret", err, reqLogger)
+		return err
+	} else if esAdminUserSecret == nil {
+		r.status.SetDegraded(operatorv1.ResourceNotFound, "Waiting for elasticsearch admin secret", nil, reqLogger)
+		return nil
+	}
+
+	// This secret should only ever contain one key.
+	if len(esAdminUserSecret.Data) != 1 {
+		r.status.SetDegraded(operatorv1.ResourceValidationError, "Elasticsearch admin user secret contains too many entries", nil, reqLogger)
+		return nil
+	}
+
+	var esAdminUserName string
+	for k := range esAdminUserSecret.Data {
+		esAdminUserName = k
+		break
+	}
+
+	// Collect the certificates we need to provision ESGW. These will have been provisioned already by the ES secrets controller.
+	cm, err := certificatemanager.Create(r.client, install, r.clusterDomain, request.TruthNamespace())
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceCreateError, "Unable to create the Tigera CA", err, reqLogger)
+		return err
+	}
+	gatewayKeyPair, err := cm.GetKeyPair(r.client, render.TigeraElasticsearchGatewaySecret, request.TruthNamespace())
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceNotFound, "Error getting TLS certificate", err, log)
+		return err
+	} else if gatewayKeyPair == nil {
+		r.status.SetDegraded(operatorv1.ResourceNotFound, "es-gateway key pair not yet available", err, log)
+		return err
+	}
+
+	kubeControllersGatewaySecret, kubeControllersVerificationSecret, kubeControllersSecureUserSecret, err := lscommon.CreateKubeControllersSecrets(ctx, esAdminUserSecret, esAdminUserName, r.client, request)
```

Comment:
This is one of those dependencies that needs to be unpicked... 

The kube-controllers secrets are created here right now because kube-controllers talks via es-gateway in order to reach ES so that it can provision users. In order to do that, it needs these secrets to exist.

The dependency chain right now looks like this:

```
es --> esgw --> kube-controllers --> linseed
```

We could probably just move the generation of these secrets into the kube-controllers function instead though - the render component for ESGW is simply installing the secrets into the cluster, so not truly needed right here.

Reply:
Is this still an accurate comment? I think you've made a comment somewhere else suggesting that we could do away with kube controllers / es gateway (in which case let's just leave this messy for now).

I'd be down for either using the ECK operator to create the necessary creds for linseed, or if that wasn't possible, just having the operator create the Elasticsearch credentials directly (and still removing needing kube controllers).

Reply:
Yeah, this has been updated now so that the operator creates the credentials for Linseed directly and there's no need for kube-controllers / esgw in multi-tenant mgmt clusters any more.

es-kube-controllers still functions as it always has for single-tenant, though.